### DISCUSSION TITLE
Get self-hosted runner name

### DISCUSSION THREAD

**Discussion by adiroiban** [community/community#25707] [Actions]
<p>Hi,</p>
<p>Is it possible to get the name of the runner on which a job is executed?</p>
<p>At the workflow yaml level or at the job / step level.</p>
<p>I have 4 bare metal servers on which I run container base tests.<br>
Each bare metal has a set of static  services… so that when a VM needs a shared service it can get it from the bare metal.</p>
<p>In the docs, I see there is only:</p>
<ul>
<li>runner.os</li>
<li>runner.temp</li>
<li>runner.tool_cache</li>
</ul>
<p><a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#runner-context" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#runner-context</a></p>

**Comment by jsoref** [comment]
<p>Technically, you could probably retrieve it from its own logs.</p>
<p>But practically, I’d just use <code>hostname</code> – that’s available on macOS, Linux, and Windows.</p>
<p>You could also probably add an extra environment variable.</p>

**Comment by adiroiban** [comment]
<p>Thanks for the reply.</p>
<p>I have this job.</p>
<p>The steps are executed inside the container… and I can’t see how to tell GitHub Action that I want a step to be executed on the docker-host / Ubuntu.</p>
<pre><code class="lang-auto">jobs:

  standard:

    name: Container
    container: ghcr.io/alpine/alpine:3.12
    runs-on: ubuntu-20.04
    timeout-minutes: 45
    steps:

    # Executed inside the container.
    - name: Alpine setup
      run: |
        apk update
        apk upgrade
        apk add shadow sudo bash git libffi

    # Executed inside the container.
    - uses: actions/checkout@v2
      with:
        clean: false

</code></pre>

**Comment by Simran-B** [comment]
<p>You could add another job <code>get-hostname</code>, run it before your current job (<code>needs: standard</code>), determine the hostname and set it as job output (<a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idoutputs" class="inline-onebox" rel="noopener nofollow ugc">Workflow syntax for GitHub Actions - GitHub Docs</a>) and then use it in the steps that are executed in the container (<code>${{ needs.get-hostname.outputs.&lt;output-name&gt; }}</code>).</p>

**Comment by airtower-luna** [comment]
<p>That second job might end up running on a different runner though, so it won’t reliably say which runner the container is on.</p>

**Comment by adiroiban** [comment, marked as answer]
<p>I ended up with this option:</p>
<p>Each docker-engine hosts bare-metal has a file with default container variables</p>
<p>/srv/docker/default.env</p>
<p>Inside that variable it can have the URL to various services</p>
<p>The /srv/docker/default.env file is managed outside of GitHub Actions/Workflow.</p>
<p>The workflow looks like this.</p>
<pre><code class="lang-auto">jobs:

  standard:

    name: Container
    container:
      image: ghcr.io/alpine/alpine:3.12
      options: --env-file /srv/docker/default.env
    runs-on: docker-host
    timeout-minutes: 45
    steps:

    # Executed inside the container.
    - name: Alpine setup
      run: |
        apk update
        apk upgrade
        apk add shadow sudo bash git libffi
        echo $DOCKER_ENGINE_HOST

    # Executed inside the container.
    - uses: actions/checkout@v2

</code></pre>
<hr>
<p>I also implemented a simple rendevous service in which each container will access an URL like <a href="http://services.local/container-services" rel="noopener nofollow ugc">http://services.local/container-services</a> and the page will return a list of service depending on the remote/source IP of the http client…</p>
<p>Basically this was a more complex version of what is my ip :slight_smile:</p>

**Comment by Simran-B** [comment]
<p>Oops, I forgot about that! Would giving each self-hosted runner a unique label and using the same machine for both jobs work? (Ideally choosing a random machine for the first job, and then picking the same for the second)</p>

**Comment by adiroiban** [comment]
<p>I don’t know about unique labels… looks complicated and I don’t understand how it works.</p>
<hr>
<p>having the runner name in the GitHub context would have been perfect,<br>
but the environment file works  fine for this case and I am happy with this solution.</p>

**Comment by acevos01** [comment]
<p><a class="mention" href="/simran-b">@Simran-B</a> Trying to make a live example from your suggestion this is the possible example:</p>
<pre><code class="lang-auto">jobs:
  job1:
    runs-on: 
      - self-hosted
      - windows
      - x64
     outputs:
      runner: ${{ steps.step1.outputs.runner }}
    steps:
      - id: step1
        run: |
            echo "::set-output name=runner::$(hostname)"
  job2:
    needs: job1
    runs-on: 
      - self-hosted
      - windows
      - x64
      - ${{ needs.job1.outputs.runner }}
    steps:
      - run: |
            echo "This was executed using: ${{ needs.dumpVars.outputs.runner }}" 
</code></pre>
